### ISSUE TITLE
drm: Provide progress messages

### ISSUE THREAD

**Issue by cmharr** [hpc/mpifileutils#195]
On large directory trees it can be hard to see how much progress is being made. Printing periodic status messages of #files deleted, deletion rate, and perhaps ETA of completion every so often (20m?) would be useful.

**Comment by adammoody** [comment]
To implement this, I'm thinking we can create a special allreduce operation that a process can call frequently and always pass in the data it would contribute to the sum used in the progress info.  This would consist of three calls:

mfu_progress_start(int secs) - this would kick off a new operation
 - rank 0: capture a timer value and save it
 - all others: start a non-blocking bcast except for rank 0

Rank 0 will start a bcast as a signal to all other procs that they need to execute an allreduce.

mfu_progress_update(int count, uint64_t values) - this would be called often, and each process would pass in the values it wants to be summed as part of the allreduce, like allreduce it takes an array of values, one special field in allreduce should be a flag that indicates whether process has more work to do or whether it is done.
 - rank 0: check whether progress timeout has expired, and if so call non-blocking bcast as signal to other ranks that they should call allreduce, call allreduce, wait on result, print result
 - all others: check whether non-blocking bcast has completed with MPI_Test, if not do nothing and exit call, if so call allreduce, then start new non-blocking bcast, then exit call.

mfu_progress_complete() - used to finalize progress messages
 - rank 0: send non-blocking bcast to start allreduce, check whether all procs are done in done flag
  -- if not, continue to send non-blocking bcast (perhaps after sleep?) with special done flag to signal time to stop, print progress result
  -- if all done, send non-blocking bcast with special done flag, print final result
 - all others: wait on non-blocking bcast
  -- if done flag is not set, send our result in allreduce (including our done flag), start a new non-blocking bcast
  -- if done flag is set, shut down

Somewhere in here, we likely need to have some user callback functions perhaps to execute the reduce operation, and certainly to print a formatted message.

**Comment by gonsie** [comment]
I'm not familiar with the code, but what does the work-loop on each rank look like? How frequently could a rank check to see if it should participate in a reduce? Also, is all the work pre-distributed after the walk... i.e., does each rank know how far along it is? 

I think we could use something asynchronous like an MPI_Ireduce or even MPI_Isend. Have all ranks send to rank 0 after a few chunks of work, and rank 0 try to poll/print every 20s or so. We don't think we need an all reduce, only rank 0 cares about global progress. Each rank could report when its 10%, 20%, 30% done... with rank 0 printing the global average periodically.

**Comment by adammoody** [comment]
Related issue: https://github.com/hpc/mpifileutils/issues/52

**Comment by adammoody** [comment]
It's better for us to use tree-based collectives to avoid flooding rank 0 with messages.  We can certainly get by with a reduce here instead of allreduce.

Although the work in most of our algorithms is static, one challenge is that different procs run through their work loops a different number of times.  That means we can't simply use the number of times through the loop as a marker on when to report, which would be an easy solution.

We could perhaps use a percent complete measure to indicate when to report, but any percentage we pick might still be too long to wait before reporting.  For example, perhaps we pick 10% but each 10% still takes an hour to finish.  So we'd want the user to be able to select this, and then it gets ugly if they start a command and then realize they picked something too big.  We'd also need to be careful to compute percentages correctly to account for procs executing enough ireduce calls, e.g., if we go with a report after every 10%, we need to correctly handle all of the cases where a process has less than 10 items.

We had a similar problem in libcircle.  We ended up going with a fixed-time based solution in libcircle, and to deal with distributed clocks, we used a bcast from rank 0 as the signal on when to do something.  Having said that, we're forced into that solution in libcircle because the work is not static like it is here.
